src-lib/yolo_layer.cpp

```diff
@@ -538,54 +553,29 @@ void *process_batch(void* ptr)
             if (mask_n >= 0) {
-                int class_id = state.truth[t * l.truth_size + b * l.truths + 4];
-                if (l.map) class_id = l.map[class_id];
-
-                int box_index = entry_index(l, b, mask_n * l.w * l.h + j * l.w + i, 0);
-                const float class_multiplier = (l.classes_multipliers) ? l.classes_multipliers[class_id] : 1.0f;
-                ious all_ious = delta_yolo_box(truth, l.output, l.biases, best_n, box_index, i, j, l.w, l.h, state.net.w, state.net.h, l.delta, (2 - truth.w * truth.h), l.w * l.h, l.iou_normalizer * class_multiplier, l.iou_loss, 1, l.max_delta, state.net.rewritten_bbox, l.new_coords);
-                (*state.net.total_bbox)++;
-
-                const int truth_in_index = t * l.truth_size + b * l.truths + 5;
-                const int track_id = state.truth[truth_in_index];
-                const int truth_out_index = b * l.n * l.w * l.h + mask_n * l.w * l.h + j * l.w + i;
-                l.labels[truth_out_index] = track_id;
-                l.class_ids[truth_out_index] = class_id;
-                //printf(" track_id = %d, t = %d, b = %d, truth_in_index = %d, truth_out_index = %d \n", track_id, t, b, truth_in_index, truth_out_index);
-
-                // range is 0 <= 1
-                args->tot_iou += all_ious.iou;
-                args->tot_iou_loss += 1 - all_ious.iou;
-                // range is -1 <= giou <= 1
-                tot_giou += all_ious.giou;
-                args->tot_giou_loss += 1 - all_ious.giou;
-
-                tot_diou += all_ious.diou;
-                tot_diou_loss += 1 - all_ious.diou;
-
-                tot_ciou += all_ious.ciou;
-                tot_ciou_loss += 1 - all_ious.ciou;
-
-                int obj_index = entry_index(l, b, mask_n * l.w * l.h + j * l.w + i, 4);
-                avg_obj += l.output[obj_index];
-                if (l.objectness_smooth) {
-                    float delta_obj = class_multiplier * l.obj_normalizer * (1 - l.output[obj_index]);
-                    if (l.delta[obj_index] == 0) l.delta[obj_index] = delta_obj;
+                if(number==0) {
+                    mloam choose = {truth, best_iou, i, j, 1, best_n, mask_n, t};
+                    mloam_ptr[number++] = choose;
+                }
+                else {
+                    int markable = 666;
+                    for(int mi = 0; mi < number; mi++) {
+                        mloam mp = mloam_ptr[mi];
+                        if(mp.best_n==best_n && mp.mask_n == mask_n && mp.x==i && mp.y==j) {
+                            markable = -666;
+                            if(mp.best_iou < best_iou) {
+                                mloam choose = {truth, best_iou, i, j, 1, best_n, mask_n, t};
+                                mloam_ptr[mi] = choose;
+                                break;
+                            }
+                        }
+                    }
+                    if(markable > 0) {
```

Comment:
Is markable a bool?  Why define it as an int and use 666 and -666?

Reply:
it is used to judge whether it is a new prediction in coordinate (x, y), if markable > 0 it means that the anchor (mask_n) in (i, j) is asigned to only one truth box in (i, j). 

if markable < 0 it means that one anchor is asigned to multi label, the anchor (mask_n) in (i, j) is asigned to more than one truth box in (i, j), so we should choose the best iou truth box. or when training, the gradient is not stable, the anchor will be asigned to truth box A, then asigned to truth box B. the same prediction in (mask_n) (i, j) will be used to predict two truth box. I think the prediction in (mask_n) (i, j) learns nothing.

Reply:
it is used to judge whether it is a new prediction in coordinate (x, y), if markable > 0 it means that the anchor (mask_n) in (i, j) is asigned to only one truth box in (i, j). 

if markable < 0 it means that one anchor is asigned to multi label, the anchor (mask_n) in (i, j) is asigned to more than one truth box in (i, j), so we should choose the best iou truth box. or when training, the gradient is not stable, the anchor will be asigned to truth box A, then asigned to truth box B. the same prediction in (mask_n) (i, j) will be used to predict two truth box. I think the prediction in (mask_n) (i, j) learns nothing.